File: ts/src/sort/selectionSort.ts
export const selectionSort = <T>(arr: T[]): T[] => {
  const arrCopy = arr.slice();

  for (let i = 0; i < arrCopy.length; i++) {

    let minIndex = i;

    for (let j = i + 1; j < arrCopy.length; j++) {
      if (arrCopy[j] < arrCopy[minIndex]) {
        minIndex = j;
      }
    }

    [arrCopy[i], arrCopy[minIndex]] = [arrCopy[minIndex], arrCopy[i]];
  }

  return arrCopy;
};

File: dataStructure/Java/src/main/java/sort/selectionSort/SelectionSort.java
package sort.selectionSort;
import sort.common.SortFunction;
import sort.common.SortUtils;

/**
 * Created by lzhao on 7/20/17.
 */

public class SelectionSort implements SortFunction {
    public int[] sort(int[] array) {
        for (int i = 0; i < array.length; i++)
        {
            int min = Integer.MAX_VALUE;
            int minIndex = -1;
            for (int j = i; j < array.length; j++)
            {
                if (array[j] < min) {
                    minIndex = j;
                    min = array[j];
                }
            }
            SortUtils.swap(array, i, minIndex);
        }

        return array;
    }
}
